src/celeritas/global/CoreState.hh

```diff
@@ -73,10 +73,23 @@ class CoreState
     //! Clear primaries after constructing initializers from them
     void clear_primaries() { num_primaries_ = 0; }
 
+    void resize_offsets(size_type n);
+
+    void count_tracks_per_action(TrackOrder order);
+
+    //! Get a range delimiting the [start, end) of the track partition assigned
+    //! action_id in track_slots
+    Range<ThreadId> get_action_range(ActionId action_id) const;
+
   private:
     // State data
     CollectionStateStore<CoreStateData, M> states_;
 
+    //
+    Collection<ThreadId, Ownership::value, M, ActionId> thread_offsets_;
+    Collection<ThreadId, Ownership::value, MemSpace::host, ActionId>
+        host_thread_offsets_;
```

Comment:
This is to avoid memory transfer every time we call `get_action_range` since we want to read the offsets on host. The problem is that this is not needed when `MemSpace` is host. Specializing `CoreState` on `MemSpace`to conditionally enable this variable doesn't seem very clean.